### PR TITLE
feat(cli): add `--output` flag to `glasskube describe`

### LINKED ISSUE #655: Add `--output` support for `glasskube describe`
See #629 for details.

### PR DESCRIPTION
**Pull request by hanshal101**
Closes #655 
Fixes #655 
Continues #673 

## 📑 Description

This PR adds the feature of getting the output in the `JSON` and `YAML` formats.

## ✅ Checks
- [ ] My code requires changes to the documentation
- [ ] I have updated the documentation as required

## ℹ Additional Information

- Output in YAML format:
```
AutoUpdate: Disabled
Configuration: {}
Dependencies: []
Entrypoints:
- Local: http://localhost:3000/
  Name: ui
  Remote: cyclops-ui:3000
LongDescription: |-
  Welcome to Cyclops, a powerful user interface for managing and interacting with Kubernetes clusters.
  Cyclops is designed to simplify the management of containerized applications on Kubernetes,
  providing an intuitive and user-friendly experience for developers, system administrators, and DevOps professionals.
  Divide the responsibility between your infrastructure and your developer teams so everyone can play to their strengths.
  Automate your processes and shrink the window for deployment mistakes.
Package: cyclops — Developer friendly Kubernetes
References:
- Label: Github
  URL: https://github.com/cyclops-ui/cyclops
- Label: Website
  URL: https://cyclops-ui.com/
- Label: Glasskube Package Manifest
  URL: https://packages.dl.glasskube.dev/packages/cyclops/v0.5.0+1/package.yaml
Repositories:
- Installed: "true"
  Name: glasskube
Status: "\e[32mReady\e[0m"
Version: 'v0.5.0+1 (latest: v0.6.0+1)'
```

- Output in JSON format:
```
{
  "AutoUpdate": "Disabled",
  "Configuration": {},
  "Dependencies": [],
  "Entrypoints": [
    {
      "Local": "http://localhost:3000/",
      "Name": "ui",
      "Remote": "cyclops-ui:3000"
    }
  ],
  "LongDescription": "Welcome to Cyclops, a powerful user interface for managing and interacting with Kubernetes clusters.\nCyclops is designed to simplify the management of containerized applications on Kubernetes,\nproviding an intuitive and user-friendly experience for developers, system administrators, and DevOps professionals.\nDivide the responsibility between your infrastructure and your developer teams so everyone can play to their strengths.\nAutomate your processes and shrink the window for deployment mistakes.",
  "Package": "cyclops — Developer friendly Kubernetes",
  "References": [
    {
      "Label": "Github",
      "URL": "https://github.com/cyclops-ui/cyclops"
    },
    {
      "Label": "Website",
      "URL": "https://cyclops-ui.com/"
    },
    {
      "Label": "Glasskube Package Manifest",
      "URL": "https://packages.dl.glasskube.dev/packages/cyclops/v0.5.0+1/package.yaml"
    }
  ],
  "Repositories": [
    {
      "Installed": "true",
      "Name": "glasskube"
    }
  ],
  "Status": "\u001b[32mReady\u001b[0m",
  "Version": "v0.5.0+1 (latest: v0.6.0+1)"
}
```

### CODE REVIEW

**Comment by hanshal101** [comment]
Description UPDATED !!!

File: cmd/glasskube/cmd/describe.go [hunk 1]
```diff
@@ -242,9 +342,60 @@ func nameAndDescription(manifest *v1alpha1.PackageManifest) string {
...
+		"Version":         version(pkg, latestVersion),
+		"Status":          status(pkg),
+		"AutoUpdate":      clientutils.AutoUpdateString(pkg, "Disabled"),
+		"Entrypoints":     entrypointsAsMap(manifest),
+		"Dependencies":    dependenciesAsMap(manifest),
+		"Repositories":    repositoriesAsMap(pkg, repos),
+		"References":      referencesAsMap(ctx, pkg, manifest),
+		"LongDescription": strings.TrimSpace(manifest.LongDescription),
+		"Configuration":   valueConfigurationsAsMap(pkg.Spec.Values),
+	}
```
Comment (kosmoz) [inline cmd/glasskube/cmd/describe.go:387]: This should definitely be refactored to a separate function which can be called from both `printYAML` and `printJSON`. Only that way can we ensure that the output schema is consistent.

Comment (kosmoz) [inline cmd/glasskube/cmd/describe.go:387]: Please start all keys with a lowercase character, since that is more common when using JSON and YAML

File: cmd/glasskube/cmd/describe.go [hunk 2]
```diff
@@ -242,9 +342,60 @@ func nameAndDescription(manifest *v1alpha1.PackageManifest) string {
...
+}
+
+func printYAML(
+	ctx context.Context,
+	pkg *v1alpha1.Package,
+	manifest *v1alpha1.PackageManifest,
+	latestVersion string,
+	repos []v1alpha1.PackageRepository) {
+	output := map[string]interface{}{
+		"Package":         nameAndDescription(manifest),
```
Comment (kosmoz) [inline cmd/glasskube/cmd/describe.go:377]: Please split this into two properties: `packageName` and `shortDescription`

File: cmd/glasskube/cmd/describe.go [hunk 3]
```diff
@@ -242,9 +342,60 @@ func nameAndDescription(manifest *v1alpha1.PackageManifest) string {
...
+		"Version":         version(pkg, latestVersion),
```
Comment (kosmoz) [inline cmd/glasskube/cmd/describe.go:378]: Please split this into three properties: `version` (taken from `pkg.Status.Version`) `desiredVersion` (taken from `pkg.Spec.PackageInfo.Version`) and `latestVersion`.

File: cmd/glasskube/cmd/describe.go [hunk 4]
```diff
@@ -242,9 +342,60 @@ func nameAndDescription(manifest *v1alpha1.PackageManifest) string {
...
+		"Status":          status(pkg),
```
Comment (kosmoz) [inline cmd/glasskube/cmd/describe.go:379]: Generally, not limited to this line: Make sure not to include any TTY color codes in the JSON/YAML output. 

In this case, please just use the `Status` field of `client.GetStatusOrPending(pkg)`.

File: cmd/glasskube/cmd/describe.go [hunk 5]
```diff
@@ -242,9 +342,60 @@ func nameAndDescription(manifest *v1alpha1.PackageManifest) string {
...
+		"AutoUpdate":      clientutils.AutoUpdateString(pkg, "Disabled"),
+		"Entrypoints":     entrypointsAsMap(manifest),
+		"Dependencies":    dependenciesAsMap(manifest),
```
Comment (kosmoz) [inline cmd/glasskube/cmd/describe.go:382]: For these two I think it would be best to just use the respective values from `manifest` directly.

File: cmd/glasskube/cmd/describe.go [hunk 6]
```diff
@@ -154,6 +206,21 @@ func printRepositories(pkg *v1alpha1.Package, repos []v1alpha1.PackageRepository
...
+func repositoriesAsMap(pkg *v1alpha1.Package, repos []v1alpha1.PackageRepository) []map[string]string {
+	repositories := []map[string]string{}
+	for _, repo := range repos {
+		repository := make(map[string]string)
+		repository["Name"] = repo.Name
+		if isInstalledFrom(pkg, repo) {
+			repository["Installed"] = "true"
+		} else {
+			repository["Installed"] = "false"
+		}
```
Comment (kosmoz) [inline cmd/glasskube/cmd/describe.go:218]: Please start the property names with lowercase characters.

Also, I think it would be better to only set this property if `isInstalledFrom` returns true (i.e. removing the `else` branch).

File: cmd/glasskube/cmd/describe.go [hunk 7]
```diff
@@ -242,9 +342,60 @@ func nameAndDescription(manifest *v1alpha1.PackageManifest) string {
...
+		"Package":         nameAndDescription(manifest),
+		"Version":         version(pkg, latestVersion),
+		"Status":          status(pkg),
+		"AutoUpdate":      clientutils.AutoUpdateString(pkg, "Disabled"),
+		"Entrypoints":     entrypointsAsMap(manifest),
+		"Dependencies":    dependenciesAsMap(manifest),
+		"Repositories":    repositoriesAsMap(pkg, repos),
+		"References":      referencesAsMap(ctx, pkg, manifest),
+		"LongDescription": strings.TrimSpace(manifest.LongDescription),
+		"Configuration":   valueConfigurationsAsMap(pkg.Spec.Values),
```
Comment (kosmoz) [inline cmd/glasskube/cmd/describe.go:386]: Also here, I think it would be better to just use `pkg.Spec.Values` directly.

**Review by kosmoz — CHANGES_REQUESTED** [review]
Hey @hanshal101! Again, sorry for taking a while to review this PR!

I talked to the rest of the team about this and we came to the conclusion that the general direction of your PR is pretty good, but we found some improvements that would make it more useful when using this feature in automations. 

Please see my comments below and let me know what you think :slightly_smiling_face:

File: cmd/glasskube/cmd/describe.go (hunk 1 above)
Comment (hanshal101) [reply cmd/glasskube/cmd/describe.go:387]: What if here I create another `struct` and use it or we need something like this:
```
func createOutputStructure(ctx context.Context, pkg *v1alpha1.Package, manifest *v1alpha1.PackageManifest, latestVersion string, repos []v1alpha1.PackageRepository) map[string]interface{} {
	return map[string]interface{}{
		"Package":         nameAndDescription(manifest),
		"Version":         version(pkg, latestVersion),
		"Status":          status(pkg),
		"AutoUpdate":      clientutils.AutoUpdateString(pkg, "Disabled"),
		"Entrypoints":     entrypointsAsMap(manifest),
		"Dependencies":    dependenciesAsMap(manifest),
		"Repositories":    repositoriesAsMap(pkg, repos),
		"References":      referencesAsMap(ctx, pkg, manifest),
		"LongDescription": strings.TrimSpace(manifest.LongDescription),
		"Configuration":   valueConfigurationsAsMap(pkg.Spec.Values),
	}
}
```

Comment (kosmoz) [reply cmd/glasskube/cmd/describe.go:387]: You can create a dedicated struct type if you prefer but I think it's not necessary. 

I would create a function like that either way.

**Comment by hanshal101** [comment]
So here my output is perfect when the package is installed. But when I try for the packages which are not installed it gives a nil pointer error.
Error:
```
goroutine 1 [running]:
github.com/glasskube/glasskube/cmd/glasskube/cmd.createOutputStructure({0x220e388, 0xc0004a1c80}, 0x0, 0xc00069e4e0, {0xc00046a520, 0xd}, {0xc0002ea640, 0x1, 0xc0005159c0?})
        H:/glasskube/cmd/glasskube/cmd/describe.go:324 +0x1fd
github.com/glasskube/glasskube/cmd/glasskube/cmd.printYAML({0x220e388?, 0xc0004a1c80?}, 0xc00000a0f0?, 0x5?, {0xc00046a520?, 0x5?}, {0xc0002ea640?, 0x0?, 0x0?})
        H:/glasskube/cmd/glasskube/cmd/describe.go:357 +0x37
github.com/glasskube/glasskube/cmd/glasskube/cmd.init.func2(0x3144a80?, {0xc00029ff50?, 0x4?, 0x1ec16ba?})
        H:/glasskube/cmd/glasskube/cmd/describe.go:75 +0x487
github.com/spf13/cobra.(*Command).execute(0x3144a80, {0xc00029fef0, 0x3, 0x3})
        C:/Users/hansh/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:987 +0xab1
github.com/spf13/cobra.(*Command).ExecuteC(0x3143f00)
        C:/Users/hansh/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1115 +0x3ff
github.com/spf13/cobra.(*Command).Execute(0x0?)
        C:/Users/hansh/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1039 +0x13
main.main()
        H:/glasskube/cmd/glasskube/main.go:11 +0x1a
```

While debugging this error I understood the `createOutputStructure` receives a nil package.
Couldn't really solve this.
NEED HELP!

**Comment by kosmoz** [comment]
> So here my output is perfect when the package is installed. But when I try for the packages which are not installed it gives a nil pointer error. Error:
> 
> ```
> goroutine 1 [running]:
> github.com/glasskube/glasskube/cmd/glasskube/cmd.createOutputStructure({0x220e388, 0xc0004a1c80}, 0x0, 0xc00069e4e0, {0xc00046a520, 0xd}, {0xc0002ea640, 0x1, 0xc0005159c0?})
>         H:/glasskube/cmd/glasskube/cmd/describe.go:324 +0x1fd
> github.com/glasskube/glasskube/cmd/glasskube/cmd.printYAML({0x220e388?, 0xc0004a1c80?}, 0xc00000a0f0?, 0x5?, {0xc00046a520?, 0x5?}, {0xc0002ea640?, 0x0?, 0x0?})
>         H:/glasskube/cmd/glasskube/cmd/describe.go:357 +0x37
> github.com/glasskube/glasskube/cmd/glasskube/cmd.init.func2(0x3144a80?, {0xc00029ff50?, 0x4?, 0x1ec16ba?})
>         H:/glasskube/cmd/glasskube/cmd/describe.go:75 +0x487
> github.com/spf13/cobra.(*Command).execute(0x3144a80, {0xc00029fef0, 0x3, 0x3})
>         C:/Users/hansh/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:987 +0xab1
> github.com/spf13/cobra.(*Command).ExecuteC(0x3143f00)
>         C:/Users/hansh/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1115 +0x3ff
> github.com/spf13/cobra.(*Command).Execute(0x0?)
>         C:/Users/hansh/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1039 +0x13
> main.main()
>         H:/glasskube/cmd/glasskube/main.go:11 +0x1a
> ```
> 
> While debugging this error I understood the `createOutputStructure` receives a nil package. Couldn't really solve this. NEED HELP!

I think the reason for the package being nil in case it is not installed is obvious, and it's the expected behavior, so the `createOutputStructure` function has to be modified such that it can handle this case. What I would do is something like this:
```go
data := map[string]any{
	"packageName":      manifest.Name,
	"shortDescription": manifest.ShortDescription,
        // ...
	// more stuff that does not depend on pkg != nil
}
if pkg != nil {
	data["desiredVersion"] = pkg.Spec.PackageInfo.Version
	// more stuff that does depend on pkg != nil
}
return data
```

But note that you don't have to do this for every occurrence of `pkg`, for example `status(pkg)` is OK, because the `status` function can already deal with `pkg == nil`.

Let me know if this helps with your problem.

**Comment by hanshal101** [comment]
Cool,, this works now !!

File: cmd/glasskube/cmd/describe.go [hunk 8]
```diff
@@ -203,14 +254,14 @@ func status(pkg *v1alpha1.Package) string {
...
+			return pkgStatus.Status
 		case string(condition.Failed):
-			return color.RedString(pkgStatus.Status)
+			return pkgStatus.Status
 		default:
 			return pkgStatus.Status
 		}
 	} else {
-		return color.New(color.Faint).Sprint("Not installed")
+		return "Not installed"
```
Comment (kosmoz) [inline cmd/glasskube/cmd/describe.go:264]: Removing the colored output here would be unfortunate. As an alternative, I would suggest not calling `status` from `createOutputStructure`, but rather setting the "status" property if `pkg != nil` like so:
```go
data["status"] = client.GetStatusOrPending(pkg).Status
```

It's OK if the status is only set if the package is installed, but optionally you can also set it to "not installed" in the initial map.

File: cmd/glasskube/cmd/describe.go [hunk 9]
```diff
@@ -176,6 +202,31 @@ func printReferences(ctx context.Context, pkg *v1alpha1.Package, manifest *v1alp
...
+		reference["url"] = ref.Url
+		references = append(references, reference)
+	}
+	if pkg != nil {
+		repo := cliutils.RepositoryClientset(ctx)
+		repoClient := repo.ForPackage(*pkg)
+		if url, err := repoClient.GetPackageManifestURL(manifest.Name, pkg.Spec.PackageInfo.Version); err == nil {
+			reference := make(map[string]string)
+			reference["Label"] = "Glasskube Package Manifest"
+			reference["URL"] = url
```
Comment (kosmoz) [inline cmd/glasskube/cmd/describe.go:223]: Please start property names with lowercase characters when converting to JSON/YAML

**Review by kosmoz — CHANGES_REQUESTED** [review]

File: cmd/glasskube/cmd/describe.go (hunk 8 above)
Comment (hanshal101) [reply cmd/glasskube/cmd/describe.go:264]: Don't we need to see the `status` even if the Package is not installed?

Comment (kosmoz) [reply cmd/glasskube/cmd/describe.go:264]: I think it would be fair to say that a missing status field would imply that a package is not installed. But as I said, you can also set the status to "not installed" in that case.

Comment (kosmoz) [inline cmd/glasskube/cmd/describe.go:264]: please revert these changes

File: cmd/glasskube/cmd/describe.go [hunk 10]
```diff
@@ -148,10 +159,25 @@ func printRepositories(pkg *v1alpha1.Package, repos []v1alpha1.PackageRepository
...
+		if isInstalledFrom(pkg, repo) {
+			repository["installed"] = "true"
 		} else {
-			fmt.Fprintln(os.Stderr)
+			repository["installed"] = "false"
 		}
+		repositories = append(repositories, repository)
 	}
+	return repositories
 }
```
Comment (kosmoz) [inline cmd/glasskube/cmd/describe.go:181]: "installed" shoud be a bool. Please change this function, so it uses
```go
repository["installed"] = isInstalledFrom(pkg, repo)
```

You will have to change the signature of the function to `[]map[string]any` for this to work.

**Review by kosmoz — CHANGES_REQUESTED** [review]
I think this PR is almost ready. Please have a look at these two things I found, then we can probably merge this.

**Review by kosmoz — APPROVED** [review]

File: cmd/glasskube/cmd/describe.go [hunk 11]
```diff
@@ -243,8 +290,66 @@ func nameAndDescription(manifest *v1alpha1.PackageManifest) string {
...
+		"dependencies":     manifest.Dependencies,
+		"longDescription":  strings.TrimSpace(manifest.LongDescription),
+		"repositories":     repositoriesAsMap(pkg, repos),
+		"references":       referencesAsMap(ctx, pkg, manifest),
+	}
+	if pkg != nil {
+		data["desiredVersion"] = pkg.Spec.PackageInfo.Version
+		data["configuration"] = pkg.Spec.Values
+		data["version"] = pkg.Status.Version
+		data["autoUpdate"] = clientutils.AutoUpdateString(pkg, "Disabled")
```
Comment (christophenne) [inline cmd/glasskube/cmd/describe.go:315]: ideally this would be a boolean instead of Enabled/Disabled – however since this PR already has quite a history, I would create a new issue for that, as this is also a quite beginner-friendly task.

**Review by christophenne — APPROVED** [review]
